In [1]:
import pandas as pd
import numpy as np
import plotly.express as px
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline


In [2]:
sns.set_style('darkgrid')
matplotlib.rcParams['font.size'] = 8
matplotlib.rcParams['figure.figsize'] = (6, 4)
matplotlib.rcParams['figure.facecolor'] = '#ffffff'

In [3]:
import warnings
warnings.filterwarnings("ignore")

In [5]:
df=pd.read_csv(r'teleCust1000t.csv')

In [6]:
df.head()

region  tenure  age  marital  address  income  ed  employ  retire  gender  \
0       2      13   44        1        9    64.0   4       5     0.0       0   
1       3      11   33        1        7   136.0   5       5     0.0       0   
2       3      68   52        1       24   116.0   1      29     0.0       1   
3       2      33   33        0       12    33.0   2       0     0.0       1   
4       2      23   30        1        9    30.0   1       2     0.0       0   

   reside  custcat  
0       2        1  
1       6        4  
2       2        3  
3       1        1  
4       4        3

In [7]:
tele_df=df.copy()

In [8]:
tele_df.head()

region  tenure  age  marital  address  income  ed  employ  retire  gender  \
0       2      13   44        1        9    64.0   4       5     0.0       0   
1       3      11   33        1        7   136.0   5       5     0.0       0   
2       3      68   52        1       24   116.0   1      29     0.0       1   
3       2      33   33        0       12    33.0   2       0     0.0       1   
4       2      23   30        1        9    30.0   1       2     0.0       0   

   reside  custcat  
0       2        1  
1       6        4  
2       2        3  
3       1        1  
4       4        3

In [9]:
tele_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 12 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   region   1000 non-null   int64  
 1   tenure   1000 non-null   int64  
 2   age      1000 non-null   int64  
 3   marital  1000 non-null   int64  
 4   address  1000 non-null   int64  
 5   income   1000 non-null   float64
 6   ed       1000 non-null   int64  
 7   employ   1000 non-null   int64  
 8   retire   1000 non-null   float64
 9   gender   1000 non-null   int64  
 10  reside   1000 non-null   int64  
 11  custcat  1000 non-null   int64  
dtypes: float64(2), int64(10)
memory usage: 93.9 KB


In [10]:
tele_df.isnull().sum()

region     0
tenure     0
age        0
marital    0
address    0
income     0
ed         0
employ     0
retire     0
gender     0
reside     0
custcat    0
dtype: int64

## EDA

In [16]:
corr_t=tele_df.corr()
corr_t

region    tenure       age   marital   address    income        ed  \
region   1.000000  0.016388  0.015034 -0.080637  0.010320  0.023765 -0.057952   
tenure   0.016388  1.000000  0.490217  0.154171  0.522952  0.242912 -0.109797   
age      0.015034  0.490217  1.000000 -0.014433  0.660435  0.327948 -0.154857   
marital -0.080637  0.154171 -0.014433  1.000000 -0.012051 -0.043984  0.001400   
address  0.010320  0.522952  0.660435 -0.012051  1.000000  0.226781 -0.128330   
income   0.023765  0.242912  0.327948 -0.043984  0.226781  1.000000  0.113939   
ed      -0.057952 -0.109797 -0.154857  0.001400 -0.128330  0.113939  1.000000   
employ   0.016335  0.520288  0.670047 -0.002891  0.463291  0.516349 -0.228580   
retire   0.005595  0.171143  0.429440 -0.078109  0.292503 -0.083651 -0.114229   
gender   0.050596  0.023533 -0.011419  0.008345  0.008363  0.038706  0.006704   
reside  -0.065158 -0.036592 -0.254993  0.625999 -0.201645 -0.097591  0.021045   
custcat -0.023771  0.166691  0.056909  0.083836  0.067913  0.134525  0.193864   

           employ    retire    gender    reside   custcat  
region   0.016335  0.005595  0.050596 -0.065158 -0.023771  
tenure   0.520288  0.171143  0.023533 -0.036592  0.166691  
age      0.670047  0.429440 -0.011419 -0.254993  0.056909  
marital -0.002891 -0.078109  0.008345  0.625999  0.083836  
address  0.463291  0.292503  0.008363 -0.201645  0.067913  
income   0.516349 -0.083651  0.038706 -0.097591  0.134525  
ed      -0.228580 -0.114229  0.006704  0.021045  0.193864  
employ   1.000000  0.277401  0.046016 -0.134683  0.110011  
retire   0.277401  1.000000  0.053906 -0.140120  0.008908  
gender   0.046016  0.053906  1.000000  0.016556 -0.004966  
reside  -0.134683 -0.140120  0.016556  1.000000  0.082022  
custcat  0.110011  0.008908 -0.004966  0.082022  1.000000

In [12]:
for i in tele_df.columns:
    print(tele_df[i].value_counts())

region
3    344
2    334
1    322
Name: count, dtype: int64
tenure
72    31
24    20
48    20
3     20
13    19
      ..
36     7
2      7
47     7
21     6
63     6
Name: count, Length: 72, dtype: int64
age
33    39
39    35
35    34
31    32
34    32
37    31
40    28
42    28
30    26
52    25
36    25
54    24
45    24
43    24
27    24
48    24
46    24
41    24
51    23
25    23
50    23
28    22
26    21
55    21
47    21
32    21
38    21
24    20
57    20
44    20
29    19
53    19
49    18
23    16
22    15
59    14
63    12
56    11
61    11
60    11
64    10
20    10
58    10
62     9
21     8
66     7
68     6
69     6
65     6
67     5
19     4
76     3
70     3
75     2
71     1
72     1
74     1
77     1
18     1
73     1
Name: count, dtype: int64
marital
0    505
1    495
Name: count, dtype: int64
address
1     68
2     66
4     61
3     61
0     56
7     53
5     50
9     41
8     39
10    38
12    36
6     36
13    33
19    29
14    24
15    24
11    23
17    22
20  

In [13]:
tele_df.head(2)

region  tenure  age  marital  address  income  ed  employ  retire  gender  \
0       2      13   44        1        9    64.0   4       5     0.0       0   
1       3      11   33        1        7   136.0   5       5     0.0       0   

   reside  custcat  
0       2        1  
1       6        4

## Chi-square test on categorical variables

In [15]:
#1.Region
import scipy.stats as stats

# Assuming 'df' is your DataFrame containing the tele_df dataset
# Selecting the 'region' and 'custcat' columns
data = tele_df[['region', 'custcat']]

# Creating a contingency table (cross-tabulation) between 'region' and 'custcat'
contingency_table = pd.crosstab(data['region'], data['custcat'])

# Performing the chi-square test
chi2_stat, p_val, dof, expected = stats.chi2_contingency(contingency_table)

# Printing the chi-square statistic, p-value, degrees of freedom, and expected frequencies
print(f"Chi-Square Statistic: {chi2_stat}")
print(f"P-Value: {p_val}")
print(f"Degrees of Freedom: {dof}")
print("Expected Frequencies:")
print(expected)

Chi-Square Statistic: 3.9005686593758746
P-Value: 0.6901301787278169
Degrees of Freedom: 6
Expected Frequencies:
[[85.652 69.874 90.482 75.992]
 [88.844 72.478 93.854 78.824]
 [91.504 74.648 96.664 81.184]]


From what we know so far, the correatlion coeff. between region and target is -0.023771(low) and the high p-value(0.6901301787278169) when conducting the chi-square test, 'region' may not have a significant association with the 'custcat' categories in the dataset. 

In [20]:
#2.Tenure
data= tele_df[['tenure','custcat']]
contingency_table=pd.crosstab(data['tenure'],data['custcat'])
chi2_stat, p_val, dof, expected = stats.chi2_contingency(contingency_table)

# Printing the chi-square statistic, p-value, degrees of freedom, and expected frequencies
print(f"Chi-Square Statistic: {chi2_stat}")
print(f"P-Value: {p_val}")
print(f"Degrees of Freedom: {dof}")
#print("Expected Frequencies:")
#print(expected)



Chi-Square Statistic: 338.2968658632468
P-Value: 9.772827270386833e-08
Degrees of Freedom: 213


The very low p-value indicates a strong association or dependency between the 'tenure' variable (representing customer tenure or duration) and the 'custcat' categories (customer segmentation categories) and we had a 0.166691 correlation coeff.

In [22]:
#3.Marital
data= tele_df[['marital','custcat']]
contingency_table=pd.crosstab(data['marital'],data['custcat'])
chi2_stat, p_val, dof, expected = stats.chi2_contingency(contingency_table)

# Printing the chi-square statistic, p-value, degrees of freedom, and expected frequencies
print(f"Chi-Square Statistic: {chi2_stat}")
print(f"P-Value: {p_val}")
print(f"Degrees of Freedom: {dof}")
#print("Expected Frequencies:")
#print(expected)

Chi-Square Statistic: 10.430746842576614
P-Value: 0.015238115782362402
Degrees of Freedom: 3


With a p-value of 0.0152, we reject the null hypothesis of independence between 'marital' and 'custcat' at a significance level of 0.05.The chi-square test indicates that marital status may have a moderate but significant influence on customer segmentation (custcat categories) in the dataset.The correlation coeff is 0.083836	








In [24]:
tele_df.columns

Index(['region', 'tenure', 'age', 'marital', 'address', 'income', 'ed',
       'employ', 'retire', 'gender', 'reside', 'custcat'],
      dtype='object')

In [28]:
#Doing a for loop might be better
cat_columns=['region', 'tenure', 'marital', 'address', 'ed',
       'employ', 'retire', 'gender', 'reside']
for i in cat_columns:
    data= tele_df[[i,'custcat']]
    contingency_table=pd.crosstab(data[i],data['custcat'])
    chi2_stat, p_val, dof, expected = stats.chi2_contingency(contingency_table)
    print(i,p_val)

    


region 0.6901301787278169
tenure 9.772827270386833e-08
marital 0.015238115782362402
address 0.41770404358762603
ed 4.815735103483139e-28
employ 0.17823160334653587
retire 0.029702819415275618
gender 0.7716993186141562
reside 0.1085697521754089


In [30]:
corr_income_age=tele_df[['age','income','custcat']].corr()
corr_income_age

age    income   custcat
age      1.000000  0.327948  0.056909
income   0.327948  1.000000  0.134525
custcat  0.056909  0.134525  1.000000

Tenure (p < 0.05), Ed (Education) (p < 0.05), Retire (p < 0.05), and Marital (p < 0.05) have p-values below the typical significance level of 0.05. This suggests significant associations or dependencies between these features and the 'custcat' categories. On the continous side, income has a weak but significant positive correaltion. Therefore , for our purposes the features we choose shall be Tenure, Education, Retire,Marital and Income

In [39]:
X=tele_df[['tenure', 'marital', 'ed',
       'retire', 'gender','income']]
y=tele_df['custcat']

## Model

In [37]:
from sklearn.model_selection import train_test_split,GridSearchCV
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.pipeline import Pipeline
from sklearn.neighbors import KNeighborsClassifier

In [44]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=42)
#Create a pipline
pipeline=Pipeline([('scaler',StandardScaler()),('classifier',KNeighborsClassifier())])
#Hyperparmeter definition
# Define hyperparameters to search over
param_grid = {
    'classifier__n_neighbors': [3, 5, 7, 10, 15],  # Number of neighbors
    'classifier__weights': ['uniform', 'distance'],  # Weighting method
    'classifier__algorithm': ['auto', 'ball_tree', 'kd_tree', 'brute'],  # Algorithm
    'classifier__leaf_size': [20, 30, 40],  # Leaf size
    'classifier__metric': ['euclidean', 'manhattan', 'chebyshev'],  # Distance metric
}

# Perform grid search with cross-validation
grid_search = GridSearchCV(pipeline, param_grid, cv=5)
grid_search.fit(X_train, y_train)

# Get the best model from the grid search
best_model = grid_search.best_estimator_

# Evaluate the best model on the test set
y_pred = best_model.predict(X_test)

# Print classification report and confusion matrix
print(classification_report(y_test, y_pred))
cnf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(cnf_matrix)

# Print the best hyperparameters found by GridSearchCV
print("Best Hyperparameters:")
print(grid_search.best_params_)



              precision    recall  f1-score   support

           1       0.40      0.43      0.42        60
           2       0.35      0.33      0.34        39
           3       0.35      0.36      0.36        55
           4       0.37      0.33      0.34        46

    accuracy                           0.37       200
   macro avg       0.37      0.36      0.37       200
weighted avg       0.37      0.37      0.37       200

Confusion Matrix:
[[26  9 16  9]
 [ 7 13 11  8]
 [17  9 20  9]
 [15  6 10 15]]
Best Hyperparameters:
{'classifier__algorithm': 'ball_tree', 'classifier__leaf_size': 20, 'classifier__metric': 'euclidean', 'classifier__n_neighbors': 7, 'classifier__weights': 'uniform'}


In [45]:
KNeighborsClassifier?
